Selective havocing in generated function bodies

xbauch opened the pull request:
Adds option `--generate-havocing-body 'fun,1,params:0;2,2,params:1` for
body-less function `fun` generates two bodies `fun.1` and `fun.2` and re-names
the respective call-sites. The first body havoc(s) the 0th and 2nd parameter,
the second body havoc(s) the 1st parameter.

The ability to selectively havoc parameter by their numbers was added to the
existing option `generate-function-body`.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 4d090089fb943532d142c0e8a0f44871e474236b).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/135982385

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 373a2a0f5e6a4f4a462bd08e0996a499a66ace05).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/135988814

hannes-steffenhagen-diffblue commented on src/goto-instrument/generate_function_bodies.h:
```diff
@@ -70,21 +70,36 @@ void generate_function_bodies(
   " --generate-function-body <regex>\n" \
   /* NOLINTNEXTLINE(whitespace/line_length) */ \
   "                              Generate bodies for functions matching regex\n" \
+  " --generate-havocing-body <option>\n" \
```
There is a `havoc` in generate function body options, is there a reason why we need a separate option for this?

hannes-steffenhagen-diffblue commented on doc/cprover-manual/goto-instrument.md:
```diff
@@ -0,0 +1,127 @@
+4th, and 5th parameter.
+
+Finally, the user can also request different implementation for a number of
+call-sites of a single function. The option `--generate-havocing-body` inserts
```
Ah, that makes more sense then! But isn’t the name `generate-havocing-body` a bit misleading then? I don’t know.

hannes-steffenhagen-diffblue reviewed (approved):
Seems reasonable to me I guess

petr-bauch committed:
Selective havocing in generated function bodies

Adds option `--generate-havocing-body 'fun,1,params:0;2,2,params:1` for
body-less function `fun` generates two bodies `fun.1` and `fun.2` and re-names
the respective call-sites. The first body havoc(s) the 0th and 2nd parameter,
the second body havoc(s) the 1st parameter.

The ability to selectively havoc parameter by their numbers was added to the
existing option `generate-function-body`.

chrisr-diffblue commented on doc/cprover-manual/goto-instrument.md:
```diff
@@ -0,0 +1,127 @@
+Which parameters should be modified can be specified either by a regular
+expression (as above) or by a semicolon-separated list of their numbers. For
+example `havoc,params:0;3;4` will assign non-deterministic values to the 1st,
+4th, and 5th parameter.
```
Is that 1st, 4th and 5th parameters regardless of type, or 1st, 4th and 5th *pointer* parameters? e.g. is it type aware or not?

xbauch replied:
So the implementation is type-safe in that it silently ignores the user-selected numbers if the respective parameter is not a pointer. Neither does it produce any error even if the user number parameter is higher than the number of parameters a given function takes. Do you think it should explicitly fail, or should we just extend the documentation?

chrisr-diffblue replied:
From an end-user point of view "silently ignore and not do what I meant" is rather annoying - so I actually think both should be done - i.e. documentation should be extended a bit, and then the tool should give an explicit error if the user has specified something that "doesn't make sense".

xbauch replied:
Done.

chrisr-diffblue reviewed (commented):
Looks reasonable, I just have a query about how typesafe this is - for instance, if the users asks to havoc parameter 1, but it turns out that parameter 1 is not a pointer type, what happens? Do they get a nice user error message, or does '1' mean 'first pointer parameter, even if there are 5 int parameters before it' ?

petr-bauch committed:
Sanity check user specified havoc parameters

1: check the function actually takes that many parameters
2: check the parameter type is a pointer

Adds tests and extends the documentation.

chrisr-diffblue reviewed (approved):
Looks good - many thanks for the additional documentation and user error messages - and for being diligent about adding additional regression tests for those new error messages :-)
